_posts/2016-05-31-unit-vs-integration-akka-testing.md

```diff
@@ -0,0 +1,143 @@
+
+
+## Unit vs Integration Testing
+
+Nowadays everybody agrees with the famous [testing pyramid](http://googletesting.blogspot.com.es/2015/04/just-say-no-to-more-end-to-end-tests.html):
+
+<img src="/assets/img/custom/blog/pyramid.png" alt="Testing pyramid" title="Testing pyramid" class="img img-center img-responsive style-screengrab">
+
+Harder to achieve is an agreement on what integration, unit, functional or acceptance means. It is reasonable as applications are structured differently depending on language, architecture and domain. I'll try to distill some essences though:
+
+* Unit: the key in here is isolation. Some people talks about isolated production code, e.g. a single function or a bunch of methods within a class (one public and the rest private). [Some other people](https://vimeo.com/68375232) talks about isolated tests, i.e. tests that may be executed on parallel as they don't access to any shared resource. A third view is that a unit test execution should be synchronous without any concurrency concern. That's [Akka's view](http://doc.akka.io/docs/akka/current/scala/testing.html):
+
+> Testing isolated pieces of code without involving the actor model, meaning without multiple threads; this implies completely deterministic behavior concerning the ordering of events and no concurrency concerns and will be called Unit Testing in the following.
+
+* Integration: this kind of tests often involves that several classes, modules or services are exercised. With Akka we'll test several actors but the key concept is that we'll use multi-threaded scheduling:
+
+> Testing (multiple) encapsulated actors including multi-threaded scheduling; this implies non-deterministic order of events but shielding from concurrency concerns by the actor model and will be called Integration Testing in the following.
+
+## Akka Unit Testing
+
+When we unit test objects we look for:
+
+* Check the returned value.
+* Verify calls to collaborators.
+* Inspect the internal state. This might be an smell on some cases.
+
+The concept of returned value is slightly different in Akka. Akka focuses on messages and not on method invocations. Checking the returned value involves two actors and two messages. If the collaborators of the *Actor Under Test* are also Actors, verifying calls will involve two actors and two messages too. If we use a multi-threaded dispatcher for scheduling this, this scenario will be out of our unit testing definition. Let's focus then on testing internal state.
```

Comment:
If we use a multi-threaded dispatcher for scheduling this, this scenario will be out of our unit testing definition.  ==> need a bit of simplification, ill point out to you more later on